src/udev/udev-node.c

```diff
@@ -232,7 +228,39 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
                                 r = -errno;
                 } while (r == -ENOENT);
 
-        return r;
+        for (i = 0; i < LINK_UPDATE_MAX_RETRIES; i++) {
+                struct stat st1 = {}, st2 = {};
+
+                r = stat(dirname, &st1);
+                if (r < 0 && errno != ENOENT)
+                        return -errno;
+
+                r = link_find_prioritized(dev, add, dirname, &target);
+                if (r < 0) {
+                        log_device_debug(dev, "No reference left, removing '%s'", slink);
+                        if (unlink(slink) == 0)
+                                (void) rmdir_parents(slink, "/");
+
+                        break;
+                } else {
+                        r = node_symlink(dev, target, slink);
+                        if (r < 0) {
+                                (void) unlink(filename);
+                                break;
+                        } else if (r == 1)
+                                continue;
+                }
+
+
+                r = stat(dirname, &st2);
+                if (r < 0 && errno != ENOENT)
+                        return -errno;
+
+                if (memcmp(&st1, &st2, sizeof(st1)) == 0)
```

Comment:
Hmm, this checks various stuff, incl. atime. And I'd expect that that will bump atime, so the test will not succeed. Shouldn't this use `stat_inode_unmodified()` instead?

Reply:
yes, please use stat_inode_unmodified()

Reply:
Hmm, I think I can't really use stat_inode_umodified because that function uses `mtime` fields of stat struct. And that is actually not a field but compatibility macro defined as `#define st_mtime st_mtim.tv_sec` hence if update happens within the same second that function would not notice the change. I could have uses memcmp() just on `st_mtim` member (struct timespec) but I was sloppy and just checked if the entire structure changed for some reason.

Reply:
Looks like `stat_inode_unmodified` should be fixed to check `st_mtim` instead.